### PR TITLE
Sensor fusion: Update of the Architecture

### PR DESCRIPTION
**Pull request by jkerpe**
- The Implementation of the Kalman Steps will be handled in the next PR. 
- Please ignore the magic numbers inside the Kalman Matrices. The values of these matrices have not yet been determined (marked with "TODO: TD124 Determine the right Parameters for LKF and EKF" inside the Code).

### CODE REVIEW

**Comment by BlueAndi** [comment]
Architecture update? Looks like the code was updated. :-)

**Comment by jkerpe** [comment]
> Architecture update? Looks like the code was updated. :-)

The name refers to the fact that these changes mainly deal with the code that is visible in e.g. a class diagram, like Method Signatures. The next PR will deal with the Implementation of the Kalman Filter Methods. Hope that made the Name a bit clear :-)

File: lib/SensorFusion/src/ExtendedKalmanFilter.h [hunk 1]
```diff
@@ -79,23 +90,91 @@ class ExtendedKalmanFilter : public LinearKalmanFilter
...
+    static const Eigen::Matrix<float, NUMBER_OF_STATES_N, NUMBER_OF_STATES_N> PROCESS_COVARIANCE_MATRIX_Q;
+
+    /** The observation model Matrix (notation in literature: H). */
+    static const Eigen::Matrix<float, NUMBER_OF_MEASUREMENTS_M, NUMBER_OF_STATES_N> OBSERVATION_MATRIX_H;
+
+    /** The covariance of the observation noise Matrix (notation in literature: R). */
+    static const Eigen::Matrix<float, NUMBER_OF_MEASUREMENTS_M, NUMBER_OF_MEASUREMENTS_M> OBSERVATION_NOISE_MATRIX_R;
+
+    /**
+     * Writes data from a KalmanParameter Struct into the Measurement Vector as a member variable m_measurementVector
```
Comment (Apoplexi) [inline lib/SensorFusion/src/ExtendedKalmanFilter.h:164]: The description does not match the naming of the method: "generate..." implies that something *new* is created, the description, however, implies that something existing is just updated.

Maybe you change the name of the method or extend the description to better clarify.

File: lib/SensorFusion/src/ExtendedKalmanFilter.h [hunk 2]
```diff
@@ -79,23 +90,91 @@ class ExtendedKalmanFilter : public LinearKalmanFilter
...
+     * with structure: [positionOdometryX, positionOdometryY, orientationOdometry]^T
+     *
+     * @param[in] kalmanParameter   Input Parameters for the Kalman Filter as a KalmanParameter struct.
+     */
+    void generateMeasurementVector(KalmanParameter& kalmanParameter);
+
+    /**
+     * Writes data from a KalmanParameter Struct into the Control Input Vector as a member variable m_controlInputVector
```
Comment (Apoplexi) [inline lib/SensorFusion/src/ExtendedKalmanFilter.h:172]: See comment above.

File: lib/SensorFusion/src/IKalmanFilter.h [hunk 3]
```diff
@@ -82,6 +83,21 @@ class IKalmanFilter
      *
      */
     virtual PositionData updateStep(KalmanParameter& kalmanParameter) = 0;
+
+private:
+    /**
+     * Writes data from a KalmanParameter Struct into the Measurement Vector as a member variable
```
Comment (Apoplexi) [inline lib/SensorFusion/src/IKalmanFilter.h:89]: See previous comment; description inconsistent with naming. (2x)

File: lib/SensorFusion/src/LinearKalmanFilter.cpp [hunk 4]
```diff
@@ -83,7 +104,17 @@ IKalmanFilter::PositionData LinearKalmanFilter::updateStep(KalmanParameter& kalm
 /******************************************************************************
  * Private Methods
  *****************************************************************************/
+void LinearKalmanFilter::generateMeasurementVector(KalmanParameter& kalmanParameter)
```
Comment (Apoplexi) [inline lib/SensorFusion/src/LinearKalmanFilter.cpp:107]: No doxygen documentation?

File: lib/SensorFusion/src/LinearKalmanFilter.h [hunk 5]
```diff
@@ -77,26 +89,86 @@ class LinearKalmanFilter : public IKalmanFilter
...
+    static const Eigen::Matrix<float, NUMBER_OF_MEASUREMENTS_M, NUMBER_OF_STATES_N> OBSERVATION_MATRIX_H;
+
+    /** The covariance of the observation noise Matrix (notation in literature: R). */
+    static const Eigen::Matrix<float, NUMBER_OF_MEASUREMENTS_M, NUMBER_OF_MEASUREMENTS_M> OBSERVATION_NOISE_MATRIX_R;
+
+    /** Initial covariance matrix (notation in literature: P). */
+    static const Eigen::Matrix<float, NUMBER_OF_STATES_N, NUMBER_OF_STATES_N> INITIAL_COVARIANCE_MATRIX_P;
+
+    /**
+     * Writes data from a KalmanParameter Struct into the measurement Vector as a member variable m_measurementVector
```
Comment (Apoplexi) [inline lib/SensorFusion/src/LinearKalmanFilter.h:158]: See previous comment; description inconsistent with naming.

File: lib/SensorFusion/src/LinearKalmanFilter.h [hunk 6]
```diff
@@ -77,26 +89,86 @@ class LinearKalmanFilter : public IKalmanFilter
...
+     * with the Structure: [positionOdometryX, positionOdometryY]
+     *
+     * @param[in] kalmanParameter   Input Parameters for the Kalman Filter as a KalmanParameter struct.
+     */
+    void generateMeasurementVector(KalmanParameter& kalmanParameter);
+
+    /**
+     * Writes data from a KalmanParameter Struct into the Control Input Vector as a member variable m_controlInputVector
```
Comment (Apoplexi) [inline lib/SensorFusion/src/LinearKalmanFilter.h:166]: See previous comment; description inconsistent with naming.

**Review by Apoplexi — CHANGES_REQUESTED** [review]

File: lib/SensorFusion/src/ExtendedKalmanFilter.cpp [hunk 7]
```diff
@@ -84,7 +107,19 @@ IKalmanFilter::PositionData ExtendedKalmanFilter::updateStep(KalmanParameter& ka
 /******************************************************************************
  * Private Methods
  *****************************************************************************/
+void ExtendedKalmanFilter::generateMeasurementVector(KalmanParameter& kalmanParameter)
```
[the remark this answers is not in the document]
Comment (jkerpe) [reply lib/SensorFusion/src/ExtendedKalmanFilter.cpp:110]: The Doxygen Documentation is only in the header-File.
Only local functions have Doxygen documentation in the cpp-File.

File: lib/SensorFusion/src/LinearKalmanFilter.cpp (hunk 4 above)
Comment (jkerpe) [reply lib/SensorFusion/src/LinearKalmanFilter.cpp:107]: The Doxygen Documentation is only in the header-File.

File: lib/SensorFusion/src/ExtendedKalmanFilter.h [hunk 8]
```diff
@@ -79,23 +90,91 @@ class ExtendedKalmanFilter : public LinearKalmanFilter
     /**
      * Initializes the variables of the Extended Kalman Filter.
      */
-    void init();
+    void init() final;
```
[the remark this answers is not in the document]
Comment (jkerpe) [reply lib/SensorFusion/src/ExtendedKalmanFilter.h:93]: Other classes are not meant to inherit from the ExtendedKalmanFilter class. Other Kalman Filter classes shall implement the Interface IKalmanFilter instead. The final specifier is used here to indicate that the functions of the Kalman Filter class cannot be overridden by derived classes.

**Review by Apoplexi — APPROVED** [review]

**Pull request merged**